### PR TITLE
Tide: Prioritize batches with pre-existing tests when picking

### PR DESCRIPTION
**Pull request by alvaroaleman**
Currently, tide will always pick the biggest possible batch. This is
suboptimal in high-traffic repos, because there is a good chance that
new PRs become merge-eligible in the timeframe it takes for tests to
fail. This means that we will have to completely start over, even if
only a single test failed in the previous attempt.

This change changes that logic to prioritize batches by number of past
successful job runs or if none exists by number of currently pending
job runs. The rationale is that:
* If we have past successful runs and assume all jobs have a non-zero
  flakerate, the likelihood of succeeding is increased
* Tide will omit the creation of jobs that already exist and are not
  failed, so this reduces the overall number of jobs we run
* If there is already an existing pending job, it is guaranteed to
  finish quicker than a new job for a new batch

The main argument against this change is that it means that we continue
testing smaller batches when a bigger one is available. In the grand
scheme of things this doesn't really make a difference though, because
if a new batch n+1 is bigger than the existing n batch, the batch
following that (n+2) will be smaller by the same number of PRs batch n+1
was bigger than n.

A nice sideeffect is that this makes reasoning easier when a subset of
jobs consistently fails (i.E. due to PRs in the batch being incompatible
or the job being broken) because one doesn't have to find the subset of
jobs that fails consistently in different batches anymore.

/cc @BenTheElder 
for thoughts on the idea

/assign @chaodaiG @cjwagner 
for thoughs and review

### CODE REVIEW

**Comment by chaodaiG** [comment]
Overall this change makes sense to me, not having enough context, aka haven't worked in k8s/kubernetes repo makes me not feeling strong about this change either. Technically this shouldn't introduce any difference in repo with zero flake rate, since there is not previous batch that didn't merge.

For repo with low flake rate, and not super high traffic, I think there is a chance that newer PR sitting in tide merge queue but not doing anything, because tide decides to still test old batch, but technically this PR has a reasonable chance of being merged in the same pool, so might be a performance hit? This might also be a problematic UX when the newer PR meets all merging requirements but tide left it not touched.

Not sure whether it's a good idea or not, I think we can do this in phases:
- Make tide merging logic configurable among the old behavior and this new behavior
- Default to old behavior, and repo(s) can opt in to this new behavior
- If it proves to work well for all, we can change the default to new

Random thoughts:
- Any chance supporting more than a single merge pool? aka a pool from old behavior and a pool from new behavior, and prefer larger pool to merge
- Tide being even smarter (I'm pretty sure we have discussed on slack before): even more merging pools

**Comment by alvaroaleman** [comment]
> For repo with low flake rate, and not super high traffic, I think there is a chance that newer PR sitting in tide merge queue but not doing anything, because tide decides to still test old batch, but technically this PR has a reasonable chance of being merged in the same pool, so might be a performance hit? 

It would delay getting that one PR merged but speed up getting something merged. Since while we finish that batch other PRs can become merge-eligible, it shouldn't really make a difference.

> This might also be a problematic UX when the newer PR meets all merging requirements but tide left it not touched.

This already happens today when we batch test, single PRs might have current, passing tests but are not going to be merged while a batch is running.

> Not sure whether it's a good idea or not, I think we can do this in phases:

Most ppl that are not intimately familiar with Prow are not going to know this exists so I am not really sure I like the idea of gating it, unless we think there might be a chance it is buggy. What would be our criteria to determine "If it proves to work well for all"?

> Any chance supporting more than a single merge pool? aka a pool from old behavior and a pool from new behavior, and prefer larger pool to merge

Supporting multiple pools could be done. However, we never trigger a batch when we already have a running batch, so I don't think that would make any difference unless we also trigger new batches even when we have a running one but that is neither trivial to implement nor to test. Additionally, this is wasteful wrt to resources, because merging one batch will validate the results of the other batch run, because it moves the basesha. This means that IMHO we shouldn't be doing this by default and only on a very explicit opt-in and at that point I'd question if its worth the added complexity.

> Tide being even smarter (I'm pretty sure we have discussed on slack before): even more merging pools

What exactly do you mean by "even more merging pools" and why is that smarter?

**Comment by stevekuznetsov** [comment]
I think I'd rather see the original plan - `log(n)` batches, each a power of two apart. Perhaps for any given size, we can prioritize re-running a batch of the same size if only one job failed. I'm not sure this particular impl has much improvement though as most jobs are going to take a similar amount of time and as long as a long-running job failed, re-running everything reduces churn on the build farm but doe not change end-user behavior.

**Comment by alvaroaleman** [comment]
> I think I'd rather see the original plan - log(n) batches, each a power of two apart. Perhaps for any given size, we can prioritize re-running a batch of the same size if only one job failed. 

Which original plan?

> I'm not sure this particular impl has much improvement though as most jobs are going to take a similar amount of time and as long as a long-running job failed, re-running everything reduces churn on the build farm but doe not change end-user behavior.

That most jobs take a similiar amount of time is speculation and often not true. If we assume a non-zero flakerate, this always increases our odds of having a success when re-triggering jobs for an existing batch if there was at least one previous success, as we now do not have a (num jobs * flake rate per job) but ((num jobs - num previous success jobs) * flake rate per job) chance.

**Comment by cjwagner** [comment]
> > Not sure whether it's a good idea or not, I think we can do this in phases:
> 
> Most ppl that are not intimately familiar with Prow are not going to know this exists so I am not really sure I like the idea of gating it, unless we think there might be a chance it is buggy. What would be our criteria to determine "If it proves to work well for all"?

I think we would want to gate this. I am only minorly worried about bugs, my main concern would be a negative impact on merge rate, specifically for the prow.k8s.io instance. Also retesting failed batch jobs will only improve merge rate if tests are actually flaking which is true in many cases, but in instances where it is not true this would hurt the merge rate since the retests won't ever succeed.
Hopefully this will actually improve the merge rate for prow.k8s.io, but since it could have a negative impact I think we should ensure that we can disable it if needed so that we don't have a code thaw crisis. Ideally this would be configurable per org or repo so that repos with healthy, non-flaky tests are not negatively impacted by retesting legitimate failures.

**Comment by cjwagner** [comment]
> > I think I'd rather see the original plan - log(n) batches, each a power of two apart. Perhaps for any given size, we can prioritize re-running a batch of the same size if only one job failed.
> 
> Which original plan?

We have previously discussed making Tide's batching logic smarter by triggering batches of varying sizes concurrently rather than a single batch at a time. In particular `log(n)` batches of size 2, 4, 8, 16, ... where `n` is the number of PRs in the pool. Concurrent batches increases the chance of a successful batch and using varying sizes balances the increased merge rate from large batches with the benefits of small batches (more likely to succeed since fewer changes correlates with fewer bugs and conflicts between PRs). When we complete a round of batch testing we'd merge the largest batch possible.

There is more to flesh out and I might have some of the details wrong, but that's what I remember.

**Comment by stevekuznetsov** [comment]
> That most jobs take a similiar amount of time is speculation and often not true. 

I mean it's true for our repos and seems to be true for k8s as well. If not these use cases, which are we building this feature for?

Screenshot from 2021-03-30 09-45-50

Ultimately I just think that it's kind of a weird feature that adds affinity and momentum to a particular batch for no real reason. Today, if a really old PR gets approved it *will* be prioritized by virtue of it's age and appear in batches. With this feature, the fact that the pool of PRs that we can batch grew 2x will not impact the size or content of a batch we trigger, which feels weird.

So I guess I'd want to think top-down - if we do not think our current batching strategy is efficient, how can we improve it? I think having this as a sub-algorithm for choosing _a specific batch size_ in a `log(n)` batch of batches might make sense, since that alleviates the downsides here, larger batches will include the PRs that would otherwise be ignored.

**Comment by alvaroaleman** [comment]
> I think we would want to gate this. I am only minorly worried about bugs, my main concern would be a negative impact on merge rate, specifically for the prow.k8s.io instance. Also retesting failed batch jobs will only improve merge rate if tests are actually flaking which is true in many cases, but in instances where it is not true this would hurt the merge rate since the retests won't ever succeed.

Why would this hurt merge rate? If we have a genuine failure, it will fail both the old and the new batch (which is guaranteed to be identical or a superset of the old one). It will also reduce the number of jobs we run, as we are not going to re-run all jobs with a new batch but just the one(s) that failed.

How would you configure the gating so that we can actually get feedback and this doesn't remain disabled forever? Or would the idea be to initially enable it but having the option to easily disable it again if there is an issue?

>> That most jobs take a similiar amount of time is speculation and often not true.

> I mean it's true for our repos and seems to be true for k8s as well. If not these use cases, which are we building this feature for?

In the screenshot you posted the quickest job took 4 min, the slowest took 1h7m so it seems to prove my point?


> Ultimately I just think that it's kind of a weird feature that adds affinity and momentum to a particular batch for no real reason. Today, if a really old PR gets approved it will be prioritized by virtue of it's age and appear in batches. With this feature, the fact that the pool of PRs that we can batch grew 2x will not impact the size or content of a batch we trigger, which feels weird.

There _is_ a very real reason: If we have jobs with a higher than zero flakechance this will improve our odds of merging something. A simple example:

* Five jobs, each have a 5% chance of flaking 
-> We have a 25% chance of at least one job failing
-> Today, we might end up starting a completely new batch, which will again have a 25% chance of at least one job failing
* With this change, the second batch triggering instead has a (5 * (5 - previous successes)% chance of failing

It is true that this can result in a smaller than possible batch being tested. However, the batch after that will be bigger by the same number of PRs the first batch was smaller. The only exception to this is if we would go above the max_batch_size limit with the second triggering attempt. I am happy to add a special case to skip this functionality if the past jobs included less than max_batch_size PRs and we have max_batch_size PRs being merge-eligible.

Also important to mention: We never start more than one batch, both before and after this PR. So if we have a repo with zero % flakerate, this will not change anything: We start a batch at some point, wait for it, merge. If there is a genuine failure in the batch, we will increase the batch size but continue to fail on the genuine failure (because we sort the PRs so the new, bigger batch is guaranteed to be a superset of the first one). In that genuine failure case we are both running jobs needlessly and make it harder to understand, because someone has to notice that the same failure in different batches occurred.

If you think any of what I wrote above is incorrect, please provide a concrete sequence of events where the behavior introduced here is worse.

>  I think having this as a sub-algorithm for choosing a specific batch size in a log(n) batch of batches might make sense, since that alleviates the downsides here, larger batches will include the PRs that would otherwise be ignored.

One extremely important point at least for openshift (not sure about kube) that seems to be completely ignored: We have limited parallel capacity due to cloud provider api quota, so we have to optimize both for running as little jobs as possible _and_ getting the highest possible merge chance. This change both reduces the number of jobs we have to run _and_ increases the odds of merging something if there is a non-zero flakerate. Triggering multiple batches of varying sizes in parallel will just result in a good chunk of them failing due to the limited api quota, so such a feature would actually need to be disabled for every openshift repo that runs e2e tests.

**Comment by stevekuznetsov** [comment]
> In the screenshot you posted the quickest job took 4 min, the slowest took 1h7m so it seems to prove my point?

No. We all know that the longer jobs flake more often. Furthermore it's clear that for the majority of jobs, if it is retriggered, the user spends at least half of the time it would have taken for the full retest of all. In many cases, they're waiting 80% of the time. I'm not saying that all jobs run the exact same duration, but that in practice this is likely to not improve merge rate.

> This change both reduces the number of jobs we have to run

I think this is a valuable way to view the content of this PR: trade-off between complexity in `tide` codebase and potential negative impact to users wondering why their batches don't work the way they expected (along with potentially stalling PRs from merging due to batch affinity) versus smaller job throughput and more efficient testing.

**Comment by BenTheElder** [comment]
Quick pass, sorry this has been sitting so long.

We have jobs that range from a few minutes to over an hour blocking kubernetes/kubernetes.

> We all know that the longer jobs flake more often

This is not strictly true. For example `pull-kubernetes-e2e-gce-100-performance` is currently at 0.99 consistency but it runs at around an hour.

> One extremely important point at least for openshift (not sure about kube) that seems to be completely ignored: We have limited parallel capacity due to cloud provider api quota, so we have to optimize both for running as little jobs as possible and getting the highest possible merge chance. 

Yes, we still do scale testing in Kubernetes presubmit for example (... another topic), which can only run with limited concurrency, so we'd have to be careful throwing lots of concurrent batches at Kubernetes.

> A nice sideeffect is that this makes reasoning easier when a subset of
jobs consistently fails (i.E. due to PRs in the batch being incompatible
or the job being broken) because one doesn't have to find the subset of
jobs that fails consistently in different batches anymore.

Yeah, it's a huge pain to track down conflicting PRs and/or flake sources in batch right now, continually adding another PR is one of the reasons.

During the last round of code freeze tide completely locked up more than once and only achieved high merge rate under human guidance kicking PRs out of the queue with `/hold`. This is very high effort with varying 15 PR batches.

> Ideally this would be configurable per org or repo so that repos with healthy, non-flaky tests are not negatively impacted by retesting legitimate failures.

Except for the conflicting PRs case (which does happen, and is hard to track down with shifting batches), surely the healthy non-flaky tests would have failed in presubmit prior to being considered for batch ...?

**Comment by stevekuznetsov** [comment]
I think at some point in the distant past we had stated that a goal of `tide` was to always make forward progress with singleton merges, so that batch merging was simply "on-top" or "free". After discussion with @alvaroaleman offline it seems like `tide` will prefer to merge the singleton instead of triggering batch tests _if and only if_ both options are available to it during one of it's sync runs every two minutes. Therefore, in principle it is true but in practice unless the timing lines up perfectly, we will always re-trigger batch testing when a singleton result is pending, so effectively `tide` will not do singleton merges while batch testing and re-testing is possible.

Originally, I thought the impact of this PR would be fairly small as any singleton merge would invalidate a batch, but as that does not seem to be the actual behavior, this PR is strictly an improvement in terms of efficiency for the phase of execution where `tide` is testing and re-testing batches.

Given any significant flake rate it *will* cause PRs that are not in the "golden" batch to wait even when eligible while re-tests happen, but as a trade-off for efficiency in testing and reduction of the total number of re-tests that need to happen for a batch to go in this can be a reasonable thing.

So to my eyes anyway we have this tradeoff:

Pros:
- more efficient testing of batches
- fewer opportunities for flakes to fail a batch

Cons:
- a little harder to understand exactly what is happening and what will happen for users
- 300LoC more in the controller logic which is already impressively hard to understand

Impact on merge throughput could go either way but can be measured as people opt in so they can make informed decisions on whether this helped or not (would depend on flake rate and their batch sizes, right?). Adding this as a feature to opt into seems reasonable.

I wonder if we can think about how we might make the core controller logic here architecturally more constrained, as each additional feature it grows only makes it harder to reason about. I'm amazed that my understanding of the "on-top" or "free" nature of `tide` no longer applied. Maybe I've just been too far away from this particular code.

**Comment by stevekuznetsov** [comment]
> a little harder to understand exactly what is happening and what will happen for users

To expand on this - it's not that this code is hard to understand or does something super complex, but I'm just thinking through the user story of a stressed-out release manager at the end of a release cycle trying to understand what's happening in the merge queue. Perhaps some UI could help here or something but it's certainly the type of question that comes up a lot - "what's the merge robot doing?!"

**Comment by alvaroaleman** [comment]
> To expand on this - it's not that this code is hard to understand or does something super complex, but I'm just thinking through the user story of a stressed-out release manager at the end of a release cycle trying to understand what's happening in the merge queue. Perhaps some UI could help here or something but it's certainly the type of question that comes up a lot - "what's the merge robot doing?!"

I agree on the general point that reasoning about Tide requires knowing an annoying amount of details. However, I think sticking to the same batch will actually make things easier for ppl trying to understand it. @BenTheElder [commment](https://github.com/kubernetes/test-infra/pull/21547#issuecomment-810529132) seems to confirm that:
> Yeah, it's a huge pain to track down conflicting PRs and/or flake sources in batch right now, continually adding another PR is one of the reasons.

**Comment by alvaroaleman** [comment]
@cjwagner @stevekuznetsov updated the PR to include a new setting for disabling this on global/org/repo level

**Review by cjwagner — APPROVED** [review]
I just have some minor nits. This LGTM.
/hold

File: prow/config/tide.go
Comment (alvaroaleman) [reply prow/config/tide.go:163]: Done

File: prow/config/tide.go
Comment (alvaroaleman) [reply prow/config/tide.go:164]: No we need the pointer to be able to differentiate "Someone explicitly set this to false" from the default bool false

File: prow/config/tide.go
Comment (alvaroaleman) [reply prow/config/tide.go]: Fixed

File: prow/config/tide.go
Comment (cjwagner) [reply prow/config/tide.go:164]: What is the distinction between an entry not existing in the map and an entry with a value of `nil` then?
If we just need to distinguish "Someone explicitly set this to false" from the default bool false I would expect `if val, ok := c.PrioritizeExistingBatchesMap[key]; ok {}` to be sufficient.
This probably wouldn't be problematic in practice, but it is currently possible to panic if a value is set to `null`.

Comment (alvaroaleman) [reply prow/config/tide.go]: Err yeah, you are ofc right and the code did actually do the ok check already in place of a nil check. Updated it to not use pointers.

**Review by cjwagner — APPROVED** [review]

**Pull request merged**
